Fix missing goto block when adding edges back to the unrolled header

a7ehuo opened the pull request:
When `newTo` is different from `newFromNextBlock` in fall into case, goto block needs to be added besides
of the cfg edges because the `newTo` block is not the fall through block.

Fixes: eclipse-openj9/openj9#15764

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

a7ehuo commented:
@hzongaro @vijaysun-omr May I ask you to review this change? Thank you! 

I'm keeping `WIP` until the Jenkins test completes. Currently the Jenkins personal build test is broken.

vijaysun-omr commented:
I feel a more descriptive explanation of the problem and how it is being fixed at a high level may be useful.

a7ehuo commented:
@vijaysun-omr I added more details to the original issue https://github.com/eclipse-openj9/openj9/issues/15764#issuecomment-1238820182 using the block that shows the issue as an example with trace logs.

a7ehuo commented on compiler/optimizer/GeneralLoopUnroller.cpp:
```diff
@@ -2489,15 +2489,92 @@ void TR_LoopUnroller::addEdgeAndFixEverything(TR_RegionStructure *region,
+                   *   | newFromNextBlockTT|  |  newToTT  |
+                   *   +-------------------+  +-----------+
+                   *
+                   *  After:
+                   *     +-----------+
+                   *     | newFromTT |
+                   *     +-----+-----+
+                   *           |
+                   *           |
+                   *     +-----v-----+
+                   *     |  gotoTT   +--------------+
+                   *     +-----+-----+              |
+                   *           |                    |
+                   *           |                    |
+                   * +---------v---------+    +-----v-----+
+                   * |newFromNextBlockTT |    |  newToTT  |
+                   * +-------------------+    +-----------+
```
I think I'll split it up into two graphs: one for the linked list of treetops, one for control flow successors.

a7ehuo replied:
Here is what I plan to update. I will wait for other comments and address them in one commit. 

```
After - TreeTop List
     +-----------+
     | newFromTT |
     +-----+-----+
           |
     +-----v-----+
     |   gotoTT  |
     +-----+-----+
           |
+----------v----------+   +---------+
|  newFromNextBlockTT |   | newToTT |
+---------------------+   +---------+

After - CFG edges
    +--------------+
    | newFromBlock |
    +-------+------+
            |
     +------v-----+       +------------+
     |  gotoBlock +-------> newToBlock |
     +------------+       +------------+

   +------------------+
   | newFromNextBlock |
   +------------------+
```

a7ehuo replied:
Addressed in ddc6a80

a7ehuo commented:
@vijaysun-omr @hzongaro Just following up and see if there are more review comments before I do another push to address [the above comment](https://github.com/eclipse/omr/pull/6693#discussion_r964876173)

hzongaro commented:
Annabelle @a7ehuo, I don't have any additional review comments.

0xdaryl commented:
@a7ehuo : are there any more pushes coming?

@hzongaro @vijaysun-omr : once Annabelle confirms or pushes final commit can you formally approve your comments?

a7ehuo commented:
> are there any more pushes coming?

Yes, I have one more commit to address [the above comment](https://github.com/eclipse/omr/pull/6693#discussion_r964876173). I wasn't sure if @vijaysun-omr would have any further comment. If not, I'll push another commit shortly

a7ehuo commented:
@vijaysun-omr @hzongaro Just pushed another commit ddc6a80 to address the comment. Ready for another review. Thanks!

hzongaro reviewed (approved):
I think the changes look good

a7ehuo commented:
I [found more information](https://github.com/eclipse-openj9/openj9/issues/15764#issuecomment-1256330901) on why https://github.com/eclipse-openj9/openj9/issues/15764 happens so far only on the modified class file. I need more time to assess if this fix is the ideal fix for this bug. Change to `WIP` status

a7ehuo commented:
@vijaysun-omr I added a [clarification](https://github.com/eclipse-openj9/openj9/issues/15764#issuecomment-1262752118) in the original issue after some offline discussion with @jdmpapin. In short, the issue hits the case where multiple blocks need to jump to the loop entry which is a fall through of the original block. This case is unexpected by `addEdgeAndFixEverything` and not common in our existing test suites. This change addresses this issue. I've removed `WIP` from this PR. It's ready for review again. Thanks!

vijaysun-omr commented:
Thanks for that clarification, it addresses my question of why this is so rare. The rare part is that we have the fall through into loop header and we are cloning the predecessor that is falling through.

vijaysun-omr commented:
Jenkins build all

a7ehuo commented:
[zos_390-64](https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/3864/console) failed with the following error. It looks like an infrastructure issue

```
10:46:40   > /u/user1/jenkins-zos/git fetch --tags --progress git@github.com:eclipse/omr.git +refs/pull/6693/merge:refs/remotes/origin/pr/6693/merge # timeout=30
10:42:58  ERROR: Error cloning remote repo 'origin'
10:42:58  hudson.plugins.git.GitException: Command "/u/user1/jenkins-zos/git fetch --tags --progress git@github.com:eclipse/omr.git +refs/pull/6693/merge:refs/remotes/origin/pr/6693/merge" returned status code 128:
10:42:58  stdout: FOTS1370 Host key verification failed.
10:42:58  fatal: Could not read from remote repository.
10:42:58  
10:42:58  Please make sure you have the correct access rights
10:42:58  and the repository exists.
10:42:58  
10:42:58  stderr: 
10:42:58  	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2675)
...
ERROR: Error cloning remote repo 'origin'
Setting status of ddc6a80d57c9c3217909b3a7d862bf191586a595 to FAILURE with url https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/3864/ and message: 'Build finished. '
Using context: continuous-integration/eclipse-omr/pr/zos_390-64
Finished: FAILURE
```

a7ehuo commented:
jenkins build zos

a7ehuo commented:
@AdamBrousseau The zos has been failing consistently with `stdout: FOTS1370 Host key verification failed` error.: [3864](https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/3864/console) and [3866](https://ci.eclipse.org/omr/job/PullRequest-zos_390-64/3866/console)  Could you help take a look?

a7ehuo commented:
The same failure was reported preivously in https://github.com/eclipse/omr/pull/6672#issuecomment-1239568705 as well
